content/blog/unsafe-string-interning-in-go.md

```diff
@@ -0,0 +1,345 @@
++++
+date = "2020-12-29T00:00:00+00:00"
+title = "Unsafe string interning in Go"
+subtitle = "I discuss the https://go4.org/intern package and the unsafe techniques it uses to implement string interning in Go."
++++
+
+* * * * * DRAFT * * * * *
+
+For a few months, I've been collaborating on the
+[`inet.af/netaddr`](https://inet.af/netaddr) package for Go which introduces a
+new `netaddr.IP` type as an alternative to the standard library's
+[`net.IP`](https://golang.org/pkg/net/#IP) type. There are [a variety of reasons
+for this](https://pkg.go.dev/inet.af/netaddr#readme-motivation), but those are
+not the focus of today's blog.
+
+The folks over at [Tailscale](https://tailscale.com/) have been working on
+improving the performance of the package in various ways, including shrinking
+the size of types and trying out faster internal IP representations.
+Specifically, one topic that came up was [how to shrink `netaddr.IP` from 32
+bytes to 24 bytes](https://github.com/inetaf/netaddr/issues/), the same size
+as a Go slice header.
+
+Because a `netaddr.IP` can also have an associated zone for IPv6 link-local
+addresses, the type previously contained a Go string, which occupies 16 bytes.
+This is a good use case for [string
+interning](https://en.wikipedia.org/wiki/String_interning), a technique which
+can be used to reduce the number of string allocations and shrink the type via a
+pointer field. It is highly likely that a small number of zone identifiers will
+be used repeatedly throughout a program, so string interning seems like a great
+fit for this problem.
+
+The result of this work is the package
+[`go4.org/intern`](https://go4.org/intern) which uses some pretty neat `unsafe`
+tricks to implement efficient string interning using weak references and Go
+finalizers. We'll start by showing off the safe implementation and gradually
+introduce the concepts needed to understand the unsafe one as well.
+
+All credit for the ideas and implementation of `go4.org/intern` goes to [Dave
```

Comment:
I would either remove me from here, or place me firmly behind Brad and Josh. I learned as much as you did about tricksy unsafe things from reading this package, my main contribution was filing a bug :)

Reply:
I alphabetized by last name but am happy to adjust as you see fit!

Reply:
Yeah I figured it was alphabetized, and normally I wouldn't mind, but showing up first for a package I didn't originate or do anything significant with feels wrong :) Unless you feel strongly, I'd just remove me from the list.

Reply:
I feel strongly. Please leave Dave in and leave it alphabetized.

Being in the room makes a difference, even if all you do is asked questions--sometimes all the more so if all you do is ask questions. And we (as a society and an industry) systematically fail to recognize and value that.

Reply:
I'll leave it as is!